encoders/firebase-encoders-processor/src/main/java/com/google/firebase/encoders/processor/getters/GetterFactory.java

```diff
@@ -0,0 +1,169 @@
+import javax.lang.model.type.DeclaredType;
+import javax.lang.model.type.ExecutableType;
+import javax.lang.model.type.TypeKind;
+import javax.lang.model.type.TypeMirror;
+import javax.lang.model.type.TypeVariable;
+import javax.lang.model.util.ElementFilter;
+import javax.lang.model.util.Elements;
+import javax.lang.model.util.Types;
+import javax.tools.Diagnostic;
+
+public final class GetterFactory {
+  private final Types types;
+  private final Elements elements;
+  private final Messager messager;
+
+  public GetterFactory(Types types, Elements elements, Messager messager) {
+    this.types = types;
+    this.elements = elements;
+    this.messager = messager;
+  }
+
+  /** Returns all getters of a given type. */
+  public Set<Getter> allGetters(DeclaredType type) {
+    if (types.isSameType(type, elements.getTypeElement("java.lang.Object").asType())) {
+      return Collections.emptySet();
+    }
+    Set<Getter> result = new LinkedHashSet<>();
+    TypeMirror superclass = ((TypeElement) types.asElement(type)).getSuperclass();
+
+    if (!superclass.getKind().equals(TypeKind.NONE)) {
+      result.addAll(allGetters((DeclaredType) superclass));
+    }
+    for (ExecutableElement method :
+        ElementFilter.methodsIn(types.asElement(type).getEnclosedElements())) {
+      create(type, method).ifPresent(result::add);
+    }
+    return result;
+  }
+
+  private Optional<Getter> create(DeclaredType ownerType, ExecutableElement element) {
+    if (element.getKind() != ElementKind.METHOD
+        || element.getModifiers().contains(Modifier.STATIC)
+        || !element.getModifiers().contains(Modifier.PUBLIC)) {
+      return Optional.empty();
+    }
+
+    ExecutableType method = (ExecutableType) element.asType();
+    if (!method.getParameterTypes().isEmpty()) {
+      return Optional.empty();
+    }
+
+    if (element.getAnnotation(Encodable.Ignore.class) != null) {
+      return Optional.empty();
+    }
+    Optional<String> fieldName = inferName(element);
+    if (!fieldName.isPresent()) {
+      return Optional.empty();
+    }
+    TypeMirror returnType = resolveTypeArguments(ownerType, element.getReturnType());
+    String getterExpression = element.toString();
+
+    TypeMirror map = types.erasure(elements.getTypeElement("java.util.Map").asType());
+    if (types.isAssignable(returnType, map)) {
+      TypeMirror keyType = ((DeclaredType) returnType).getTypeArguments().get(0);
+      if (!types.isSameType(keyType, elements.getTypeElement("java.lang.String").asType())) {
+        messager.printMessage(
+            Diagnostic.Kind.ERROR,
+            "Cannot encode Maps with non-String keys.",
+            ((DeclaredType) returnType).asElement());
+      }
+    }
+    if (types.isAssignable(
+        returnType, types.erasure(elements.getTypeElement("java.util.Optional").asType()))) {
+      returnType = ((DeclaredType) returnType).getTypeArguments().get(0);
+      getterExpression = getterExpression + ".orElse(null)";
+    }
```

Comment:
How likely is that we need to support optional? I though it wasn't available in android in anything less than API24

Reply:
It's available in google3 and will be supported by agp 4 desugaring.

Reply:
lgtm